be/src/util/arrow/starrocks_column_to_arrow.cpp

```diff
@@ -236,12 +253,262 @@ struct ColumnToArrowConverter<LT, AT, is_nullable, ConvBinaryGuard<LT, AT>> {
+        auto& element_arrow_type = down_cast<arrow::ListType*>(column_context->arrow_type.get())->field(0)->type();
+        auto func = resolve_convert_func(element_type_desc.type, element_arrow_type->id(), element_is_nullable);
+        if (func == nullptr) {
+            return arrow::Status::NotImplemented(fmt::format(
+                    "Can't find convert function for array element, logical type: {}, arrow type: {}, nullable: {}",
+                    type_to_string(element_type_desc.type), element_arrow_type->name(), element_is_nullable));
+        }
+        column_context->child_column_contexts.push_back({element_type_desc, element_arrow_type, func});
+        return arrow::Status::OK();
+    }
+
+    static inline arrow::Status convert(const ColumnPtr& column, int start_idx, int end_idx,
+                                        ColumnContext* column_context, arrow::ArrayBuilder* array_builder) {
+        ARROW_RETURN_NOT_OK(check_const_and_only_null(column));
+        ARROW_RETURN_NOT_OK(initialize_child_column_context(column, column_context));
+        auto& element_column_context = column_context->child_column_contexts[0];
+        arrow::ListBuilder* builder = down_cast<arrow::ListBuilder*>(array_builder);
+        auto element_builder = builder->value_builder();
+        if constexpr (is_nullable) {
+            const auto* nullable_column = down_cast<NullableColumn*>(column.get());
+            const auto* data_column = down_cast<ArrayColumn*>(nullable_column->data_column().get());
+            auto& offsets = data_column->offsets().get_data();
+            const auto& element_column = data_column->elements_column();
+            ARROW_RETURN_NOT_OK(element_builder->Reserve(end_idx - start_idx + 1));
+            for (auto i = start_idx; i <= end_idx; ++i) {
+                if (nullable_column->is_null(i)) {
+                    ARROW_RETURN_NOT_OK(builder->AppendNull());
+                } else {
+                    ARROW_RETURN_NOT_OK(builder->Append());
+                    ARROW_RETURN_NOT_OK(element_column_context.convert_func(
+                            element_column, offsets[i], offsets[i + 1] - 1, &element_column_context, element_builder));
+                }
+            }
+        } else {
+            const auto* data_column = down_cast<ArrayColumn*>(column.get());
+            auto& offsets = data_column->offsets().get_data();
+            const auto& child_column = data_column->elements_column();
+            ARROW_RETURN_NOT_OK(element_builder->Reserve(end_idx - start_idx + 1));
+            for (auto i = start_idx; i <= end_idx; ++i) {
+                ARROW_RETURN_NOT_OK(builder->Append());
+                ARROW_RETURN_NOT_OK(element_column_context.convert_func(child_column, offsets[i], offsets[i + 1] - 1,
+                                                                        &element_column_context, element_builder));
+            }
+        }
```

Comment:
why not direct nest call convert nullable/non-column conver?

Reply:
According to arrow [BaseListBuilder](https://github.com/apache/arrow/blob/release-5.0.0/cpp/src/arrow/array/builder_nested.h#L42)  API, there are two ways to build the list
option 1. first call `AppendValues` to set offsets and null bitmap in batch, then nest call converter to build value
option 2. for each array, first call `Append`/`AppendNull`, then nest call converter to build values for this array
I think you mean why not use option 1.

option 1 is more efficient but has some limitations according to the `AppendValues` parameters
* `const offset_type* offsets`: the offsets should be calculated in advance so that `AppendValues` can just do a memcpy. If `start_idx` = 0, we can directly use `ArrayColumn::_offsets` as the parameter
* `const uint8_t* valid_bytes`: any **zero byte** will be considered as a null for that slot, and also should be calculated in advance. Here we can not use `NullColumn(FixedLengthColumn<uint8_t>)` buffer  as the parameter because in our system **non zero byte** will be considered as a null. We need to reverse the byte from 1 to 0 / 0 to 1
Currently I do not reverse the null byte to use option 1 because it also has some cost, and I can not decide the benefit.